cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java

```diff
@@ -0,0 +1,228 @@
+    private final long timestamp;
+    private final TransactionVisibilityState state;
+
+    private byte[] prevTopicIdBytes;
+    private Long currentTTL;
+    private Integer currentGen;
+
+    MessageDataFilter(RegionCoprocessorEnvironment env, long timestamp) {
+      this.env = env;
+      this.timestamp = timestamp;
+      this.state = txStateCache.getLatestState();
+    }
+
+    @Override
+    public ReturnCode filterKeyValue(Cell cell) throws IOException {
+      int rowKeyOffset = cell.getRowOffset() + prefixLength;
+      int sizeOfRowKey = cell.getRowLength() - prefixLength;
+      long publishTimestamp = MessagingUtils.getPublishTimestamp(cell.getRowArray(), rowKeyOffset, sizeOfRowKey);
+      int topicIdLength = MessagingUtils.getTopicLengthMessageEntry(sizeOfRowKey) - Bytes.SIZEOF_INT;
+      int generationId = Bytes.toInt(cell.getRowArray(), rowKeyOffset + topicIdLength);
+
+      try {
+        if (prevTopicIdBytes == null || currentTTL == null || currentGen == null ||
+          (!org.apache.hadoop.hbase.util.Bytes.equals(prevTopicIdBytes, 0, prevTopicIdBytes.length,
+                                                      cell.getRowArray(), rowKeyOffset, topicIdLength))) {
+          prevTopicIdBytes = Arrays.copyOfRange(cell.getRowArray(), rowKeyOffset, rowKeyOffset + topicIdLength);
+          Map<String, String> properties = topicCache.get(ByteBuffer.wrap(prevTopicIdBytes));
+          currentTTL = Long.parseLong(properties.get(MessagingUtils.Constants.TTL_KEY));
+          currentGen = Integer.parseInt(properties.get(MessagingUtils.Constants.GENERATION_KEY));
+        }
+
+        // Old Generation (or deleted current generation) cleanup
+        if (MessagingUtils.isOlderGeneration(generationId, currentGen)) {
+          return ReturnCode.SKIP;
+        }
+
+        // TTL expiration cleanup only if the generation of metadata and row key are the same
+        if ((generationId == Math.abs(currentGen)) && ((timestamp - publishTimestamp) > currentTTL)) {
+          return ReturnCode.SKIP;
+        }
+      } catch (ExecutionException ex) {
+        LOG.debug("Region " + env.getRegion().getRegionNameAsString() + ", exception while" +
+                    "trying to fetch properties of topicId {}" + MessagingUtils.toTopicId(prevTopicIdBytes), ex);
+      }
+      return ReturnCode.INCLUDE;
+    }
+
+    @Override
+    public Cell transformCell(Cell cell) throws IOException {
+      // Rollback expired/invalidated transactions
+      byte[] txCol = MessagingUtils.Constants.TX_COL;
+      if (org.apache.hadoop.hbase.util.Bytes.equals(cell.getQualifierArray(), cell.getQualifierOffset(),
+                                                    cell.getQualifierLength(), txCol, 0, txCol.length)) {
+        // This is a TX COL
+        long txWritePtr = Bytes.toLong(cell.getValueArray(), cell.getValueOffset());
+        if (txWritePtr > 0 && state != null && state.getInvalid().contains(txWritePtr)) {
```

Comment:
It is actually a list. So you can do a small optimization by first check if it is instance of `List`. If it is, then cast and use `Collections.binarySearch`, otherwise fallback to `contains` method.